book/src/internals/type-system.md

```diff
@@ -0,0 +1,224 @@
+>
+> -- Wikipedia
+
+## Purpose
+
+Each of the SQL DBMSs has their own type system. Thanks to SQL standard, they
+are very similar, but have key differences. For example, SQLite does not have a
+type for date or time or timestamps, but it has functions for handling date and
+time that take ISO8603 strings or ints representing Unix timestamps. So it does
+support most of what is possible to do with dates, it just stores a different
+data format and uses different functions to achieve that.
+
+PRQL's task is to define common description of _data formats_, just as how it
+already defines _data transformations_.
+
+I believe this should best be done in two steps:
+
+1. Define PRQL's Type System (PTS), following principles we think a relational
+   language should have (and not focus on what existing SQL DBMSs have).
+
+2. Define a mapping from the SQL's Type System (STS) into PTS, for each of the
+   DBMSs. Ideally we'd want that to be a bijection, so each type in PTS would be
+   represented by a single type in STS and vice-versa. Unfortunately this is not
+   entirely possible, as shown below.
+
+In practical terms, we want for a user to be able to:
+
+- ... express types of their database with PRQL. In some cases, we can allow to
+  say "your database is not representable with PRQL, change it or use only a
+  subset of it". An example of what we don't want to support are arrays with
+  arbitrary indexes in Postgres (i.e. 2-based index for arrays). This task could
+  be automated by LSP server, via introspecting user's database and producing
+  PRQL source.
+
+- ... express their SQL queries in PRQL.
+
+- ... use any PRQL feature in their database. For example, translate PRQL's
+  datetime operations to use TEXT in SQLite.
+
+Example of the mapping between PTS and two STSs:
+
+| PTS       | STS Postgres | STS SQLite |
+| --------- | ------------ | ---------- |
+| int32     | integer      | INTEGER    |
+| int64     | bigint       | INTEGER    |
+| timestamp | timestamp    | TEXT       |
+
+## Principles
+
+**Algebraic types** - have a way of expressing sum and product types. In Rust,
+sum would be an enum and product would be tuple or a struct. In SQL, product
+would be a row, since it can contain different types, all at once. Sum would be
+harder to express, see (this
+post)[https://www.parsonsmatt.org/2019/03/19/sum_types_in_sql.html ].
+
+The value proposition here is that algebraic types give a lot modeling
+flexibility, all while being conceptually simple.
+
+**Composable** - as with transformation, we'd want types to compose together.
+
+Using Python, JavaScript, C++ or Rust, one could define many different a data
+structure that would correspond to our idea of "relation". Most of them would be
+an object/struct that has column names and types and then a generic array of
+arrays for rows.
+
+PRQL's type system should also be able to express relation as composed from
```

Comment:
relation^s^

Reply:
@snth easy to add suggestions and then the author can just click

```suggestion
PRQL's type system should also be able to express relations as composed from
```

Reply:
Thanks @max-sixty . I didn't know that. Will try and do that from now on.

Reply:
Couldn't figure this out on my own but got it now thanks to this: https://egghead.io/lessons/github-add-suggestions-in-a-github-pr-review